subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/swift/SwiftcToolChain.java

```diff
@@ -0,0 +1,120 @@
+    private final Instantiator instantiator;
+    private final ToolSearchPath toolSearchPath;
+    private final ExecActionFactory execActionFactory;
+
+    public SwiftcToolChain(String name, BuildOperationExecutor buildOperationExecutor, OperatingSystem operatingSystem, PathToFileResolver fileResolver, ExecActionFactory execActionFactory,
+                           CompilerOutputFileNamingSchemeFactory compilerOutputFileNamingSchemeFactory, CompilerMetaDataProviderFactory compilerMetaDataProviderFactory, Instantiator instantiator) {
+        this(name, buildOperationExecutor, operatingSystem, fileResolver, execActionFactory, compilerOutputFileNamingSchemeFactory, new ToolSearchPath(operatingSystem), compilerMetaDataProviderFactory, instantiator);
+    }
+
+    SwiftcToolChain(String name, BuildOperationExecutor buildOperationExecutor, OperatingSystem operatingSystem, PathToFileResolver fileResolver, ExecActionFactory execActionFactory,
+                    CompilerOutputFileNamingSchemeFactory compilerOutputFileNamingSchemeFactory, ToolSearchPath tools, CompilerMetaDataProviderFactory compilerMetaDataProviderFactory, Instantiator instantiator) {
+        super(name, buildOperationExecutor, operatingSystem, fileResolver);
+        this.compilerOutputFileNamingSchemeFactory = compilerOutputFileNamingSchemeFactory;
+        this.instantiator = instantiator;
+        this.toolSearchPath = tools;
+        this.execActionFactory = execActionFactory;
+    }
+
+    @Override
+    public List<File> getPath() {
+        return toolSearchPath.getPath();
+    }
+
+    @Override
+    public void path(Object... pathEntries) {
+        for (Object path : pathEntries) {
+            toolSearchPath.path(resolve(path));
+        }
+    }
+
+    @Override
+    public PlatformToolProvider select(NativePlatformInternal targetPlatform) {
+        return new AbstractPlatformToolProvider(buildOperationExecutor, targetPlatform.getOperatingSystem()) {
+            @Override
+            public <T extends CompileSpec> org.gradle.language.base.internal.compile.Compiler<T> newCompiler(Class<T> spec) {
+                if (SwiftCompileSpec.class.isAssignableFrom(spec)) {
+                    return CompilerUtil.castCompiler(createSwiftCompiler());
+                }
+                return super.newCompiler(spec);
+            }
+
+            protected Compiler<SwiftCompileSpec> createSwiftCompiler() {
+                GccCommandLineToolConfigurationInternal swiftCompilerTool = instantiator.newInstance(DefaultGccCommandLineToolConfiguration.class, ToolType.SWIFT_COMPILER, "swiftc");
+                SwiftCompiler swiftCompiler = new SwiftCompiler(buildOperationExecutor, compilerOutputFileNamingSchemeFactory, commandLineTool(swiftCompilerTool), context(swiftCompilerTool), getObjectFileExtension(), false);
+                return new OutputCleaningCompiler<SwiftCompileSpec>(swiftCompiler, compilerOutputFileNamingSchemeFactory, getObjectFileExtension());
```

Comment:
I don't think we need the output cleaning wrapper, if we don't have separate compile and link steps.